pytorch_lightning/trainer/lr_finder.py

```diff
@@ -0,0 +1,422 @@
+"""
+Trainer Learning Rate Finder
+"""
+from abc import ABC, abstractmethod
+from typing import Optional
+
+import numpy as np
+import torch
+from torch.optim.lr_scheduler import _LRScheduler
+from torch.utils.data import DataLoader
+from tqdm.auto import tqdm
+import os
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.callbacks import Callback
+from pytorch_lightning import _logger as log
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class TrainerLRFinderMixin(ABC):
+    @abstractmethod
+    def _atomic_save(self, *args):
+        """Warning: this is just empty shell for code implemented in other class."""
+
+    def _run_lr_finder_internally(self, model):
+        """ Call lr finder internally during Trainer.fit() """
+        lr_finder = self.find_lr(model)
+        lr = lr_finder.suggestion()
+        log.info(f'Learning rate set to {lr}')
+        if isinstance(self.auto_lr_find, str):
+            if hasattr(model.hparams, self.auto_lr_find):
+                setattr(model.hparams, self.auto_lr_find, lr)
+            else:
+                raise MisconfigurationException(
+                    f'`auto_lr_find` was set to {self.auto_lr_find}, however'
+                    ' could not find this as a field in model.hparams.')
+        else:
+            if hasattr(model.hparams, 'lr'):
+                model.hparams.lr = lr
+            elif hasattr(model.hparams, 'learning_rate'):
+                model.hparams.learning_rate = lr
+            else:
+                raise MisconfigurationException(
+                    'When auto_lr_find is set to True, expects that hparams'
+                    ' either has field `lr` or `learning_rate` that can overridden')
+
+    def _model_dump(self, filepath, model):
```

Comment:
is there a reason we can't use the Trainer's `save_checkpoint` and `restore` here?

Reply:
Only problem I found is that `trainer.optimizer` and `trainer.scheduler` is not initialized at this point and the standard `save_checkpoint` therefore does not work. Could probably take care of that by setting them to an empty list. Will try to change to existing method. Thanks :+1: